### PR TITLE
Add `--no-header` option

### LINKED ISSUE #4998: Suppress `godot` output, but not script output with `-q`
### Describe the project you are working on

Command line script that uses Godot for visualization and stdout for piping results.

### Describe the problem or limitation you are having in your project

`stdout` from script is always prepended with Godot version and debug info.
```
Godot Engine v3.4.4.stable.fedora.419e713a2 - https://godotengine.org
OpenGL ES 3.0 Renderer: Mesa Intel(R) HD Graphics 4400 (HSW GT2)
OpenGL ES Batching: ON
```
Using `--quiet` removes all `stdout` messages, including those from script,
and I'd to still read script messages.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Allow to specify `--quiet` (or `-q` in 4.0) multiple times.

`-q` - silences `godot` messages, but leaves script output
`-q -q` - silences all `stdout` output

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Pseudo-code.

https://github.com/godotengine/godot/pull/63575/files

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Not from within Godot.

### Is there a reason why this should be core and not an add-on in the asset library?

Addons don't have control over internal Godot messages.

### PR DESCRIPTION
**Pull request by abitrolly**
The option to suppress printing Godot headers, see https://github.com/godotengine/godot-proposals/issues/4998 for details.

Closes https://github.com/godotengine/godot-proposals/issues/4998.
Closes #63575.

### CODE REVIEW

**Review by ghost — CHANGES_REQUESTED** [review]
To add this feature to the project settings, I suggest you add the following lines of code:

in core/config/project_settings.cpp to line <b>1200</b>:
```
GLOBAL_DEF("application/run/quit_header", false); 
```

and in main.cpp to line <b>1600</b> :
```
if (bool(GLOBAL_GET("application/run/quit_header"))){
		Engine::singleton->_print_header = false;
}
```

that's about it!

**Comment by Calinou** [comment]
> GLOBAL_DEF("application/run/quit_header", false); 

Why is the setting called `quit_header` instead of `print_header`? Is that a typo?

**Comment by kaissouDev** [comment]
yes, it's a mistake on my part, but you can now call the settings "disable header".

**Comment by kaissouDev** [comment]
I'll test it and let you know if it works when I've finished running my workflow.

**Comment by kaissouDev** [comment]
I tested it with the project parameters and it worked!

**Comment by abitrolly** [comment]
@kaissouDev I'd prefer to see this PR reviewed/merged first, and then address project setting logic in a different PR. To keep it simple.

**Comment by kaissouDev** [comment]
Yes, you're right, it's better to wait until this pull request is merged. At least we know how to add this option to the project parameters.

**Comment by abitrolly** [comment]
Rebased, solved conflicts. Testing with [the hello script](https://docs.godotengine.org/en/stable/tutorials/editor/command_line_tutorial.html#running-a-script).

```bash
$ /bin/godot.linuxbsd.editor.x86_64 -s hello.gd
Godot Engine v4.3.dev.custom_build.7bfed92d9 - https://godotengine.org
Inconsistent value (1) for DRI_PRIME. Should be < 1 (GPU devices count). Using: 0
OpenGL API 4.6 (Core Profile) Mesa 23.3.5 - Compatibility - Using Device: Intel - Mesa Intel(R) HD Graphics 4400 (HSW GT2)
 
Hello!
Project is missing: /tmp/ggg/project.godot 
```

```bash
$ ./bin/godot.linuxbsd.editor.x86_64 -s hello.gd --no-header
Inconsistent value (1) for DRI_PRIME. Should be < 1 (GPU devices count). Using: 0
Hello!
Project is missing: /tmp/ggg/project.godot
```

I will now try to add this to projects settings too.

**Comment by abitrolly** [comment]
This is now ready for review.

@Calinou @kaissouDev

File: doc/classes/ProjectSettings.xml
Comment (Mickeon) [inline doc/classes/ProjectSettings.xml:321]: It's more common in the class reference to talk about "What happens if **enabled**", regardless of the default value. It's also not a bad idea to be more explicit here.
```suggestion
			If set to [code]true[/code], the engine header is printed in the console on startup. This header describes the current version of the engine, as well as the renderer being used.
```

**Review by Mickeon — APPROVED** [review]
Docs do look good now. Remember to [squash your commits](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase) now. Thank you for your patience very much

File: core/config/engine.h
Comment (AThousandShips) [inline core/config/engine.h:128]:
```suggestion
	void print_header(const String &p_string) const;
```

File: core/config/engine.cpp
Comment (AThousandShips) [inline core/config/engine.cpp:261]:
```suggestion
void Engine::print_header(const String &p_string) const {
```

**Review by AThousandShips — APPROVED** [review]

**Comment by abitrolly** [comment]
Squashed into a single commit.

**Comment by abitrolly** [comment]
@AThousandShips what is the `p_string` convention for function argument? I just followed the style, but it is still interesting what `p_` stands for.

**Comment by AThousandShips** [comment]
"parameter" :)

File: main/main.cpp
Comment (Calinou) [inline main/main.cpp:486]:
```suggestion
	print_help_option("--no-header", "Do not print engine version and rendering method header on startup.\n");
```

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected. Should be good to merge after applying the suggestion and squashing commits together.

**Comment by abitrolly** [comment]
@Calinou the rendering method info strings are gone from Vulkan backend after rebasing over https://github.com/godotengine/godot/pull/87340 but that can be added later if needed. So this is now ready.

**Comment by Calinou** [comment]
> @Calinou the rendering method info strings are gone from Vulkan backend after rebasing over #87340 but that can be added later if needed. So this is now ready.

I noticed this while testing – OpenGL header is silenced, but not Vulkan header. Could you look into resolving this?

**Comment by abitrolly** [comment]
@Calinou can you give me the message? I don't know what to look for..

**Comment by Calinou** [comment]
> @Calinou can you give me the message? I don't know what to look for..

The message is here: https://github.com/godotengine/godot/blob/907db8eebcecb97d527edcaff77a1c87a6c068f5/servers/rendering/rendering_device.cpp#L4986

**Comment by abitrolly** [comment]
@Calinou done.

File: doc/classes/ProjectSettings.xml
Comment (akien-mga) [inline doc/classes/ProjectSettings.xml:357]:
```suggestion
			If [code]true[/code], the engine header is printed in the console on startup. This header describes the current version of the engine, as well as the renderer being used. This behavior can also be disabled on the command line with the [code]--no-header[/code] option.
```

File: main/main.cpp
Comment (akien-mga) [inline main/main.cpp:1018]: Both seem to work, but below you're using this form, it would be best to keep it consistent:
```suggestion
			Engine::get_singleton()->_print_header = false;
```

**Review by akien-mga — APPROVED** [review]
Looks good to me. Last few nitpicks :)

I noticed that @AThousandShips is credited as co-author twice in your commit message, you could amend it to remove the duplicate line. (Not critical though, but this could be done while doing the other suggested changes.)

**Comment by abitrolly** [comment]
@akien-mga done.

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Comment by kaissouDev** [comment]
Good job @abitrolly !

**Pull request merged**
